File: usaco/train/ch1/1.2_adhoc/gift1.java
/* 
ID: aaryan.1
PROG: gift1
LANG: JAVA
*/

import java.util.*;
import java.io.*;

public class gift1 {
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new FileReader("gift1.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("gift1.out")));
        int NP = Integer.parseInt(in.readLine()); 
        String[] names = new String[NP];
        int[] money = new int[NP];
        for (int i = 0; i < NP; i++) {
            names[i] = in.readLine(); 
        }

        String giver; 
        StringTokenizer st;
        int amt;
        int NG;

        for (int j = 0; j < NP; j++) {
            giver = in.readLine();
            st = new StringTokenizer(in.readLine());
            amt = Integer.parseInt(st.nextToken());
            NG = Integer.parseInt(st.nextToken());
            if (NG != 0) {
                money[find(names, giver)] = money[find(names, giver)] - amt + (amt % NG);
                for (int i = 0; i < NG; i++) {
                    String n = in.readLine(); 
                    money[find(names, n)] += (amt / NG); 
                } 
            }
        }
        
        for (int i = 0; i < NP; i++) {
            out.println(names[i] + " " + money[i]);
        }

        out.close();
    }

    public static int find(String[] names, String name) {
        for (int i = 0; i < names.length; i++) {
            if (names[i].equals(name)) {
                return i; 
            }
        }

        return -1; 
    }
}

File: usaco/train/ch1/1.2_adhoc/gift1.cpp
/*
ID: aaryan.1
PROB: gift1
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <string> 

using namespace std; 

int find(vector <string> &names, string name) {
    for (int i = 0; i < names.size(); i++) {
        if (names[i].compare(name) == 0) {
            return i; 
        }
    }

    return -1; 
}

int main() {
    ifstream fin ("gift1.in");
    //ofstream fout ("gift1.out");
    int NP; fin >> NP; 
    vector <string> names;
    for (int i = 0; i < NP; i++) {
        string name; fin >> name;
        names.push_back(name);
    } 

    int money[NP];

    string giver; fin >> giver; 
    int amt, NG; fin >> amt >> NG; 

    while (amt != 0 && NG != 0) {
        money[find(names, giver)] = money[find(names, giver)] - amt + (amt % NG);
        for (int i = 0; i < NG; i++) {
            string n; fin >> n;
            money[find(names, n)] += amt / NG;  
        }
    }

    for (int i = 0; i < NP; i++) {
        cout << names[i] + " " + money[i] << endl; 
    }
}